Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,104 @@
+        try swizzler.unswizzle(selector: sel_DataTaskWithRequest, in: subjectClass)
+        try swizzler.unswizzle(selector: sel_DataTaskWithURLCompletion, in: subjectClass)
+        try swizzler.unswizzle(selector: sel_DataTaskWithRequestCompletion, in: subjectClass)
+    }
+
+    static func swizzleDataTaskWithURL() throws {
+        // typealiases cannot be generic as C/block conventions don't support generics
+        // note that _Block doesn't have Selector parameter
+        typealias TypedIMP = DataTaskWithURL_C
+        typealias TypedBlockIMP = DataTaskWithURL_Block
+
+        let sel = sel_DataTaskWithURL
+        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+
+        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURL -> URLSessionDataTask in
+            return typedOriginalImp(impSelf, impSelector, impURL)
+        }
+        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
+        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
+    }
+
+    static func swizzleDataTaskWithRequest() throws {
+        typealias TypedIMP = DataTaskWithRequest_C
+        typealias TypedBlockIMP = DataTaskWithRequest_Block
+
+        let sel = sel_DataTaskWithRequest
+        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+
+        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURLRequest -> URLSessionDataTask in
+            return typedOriginalImp(impSelf, impSelector, impURLRequest)
+        }
+        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
+        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
+    }
+
+    static func swizzleDataTaskWithURLCompletionHandler() throws {
+        typealias TypedIMP = DataTaskWithURLCompletion_C
+        typealias TypedBlockIMP = DataTaskWithURLCompletion_Block
+
+        let sel = sel_DataTaskWithURLCompletion
+        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+
+        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURL, impCompletion -> URLSessionDataTask in
+            return typedOriginalImp(impSelf, impSelector, impURL, impCompletion)
+        }
+        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
+        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
+    }
+
+    static func swizzleDataTaskWithRequestCompletionHandler() throws {
+        typealias TypedIMP = DataTaskWithRequestCompletion_C
+        typealias TypedBlockIMP = DataTaskWithRequestCompletion_Block
+
+        let sel = sel_DataTaskWithRequestCompletion
+        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+
+        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURLRequest, impCompletion -> URLSessionDataTask in
```

Comment:
Don't we need also `URLSessionDataTaskSwizzler` which will allow us intercepting `resume()` and `cancel()` calls?

Reply:
yes, if we could KVO it would be trivial but it seems like KVO in Swift is broken due to threading issues.

i'm still trying to figure out how we can do that as `cancel/suspend` don't always get called as far as i see